allenact/utils/experiment_utils.py

```diff
@@ -408,8 +458,57 @@ class OffPolicyPipelineComponent(NamedTuple):
     ] = lambda cur_worker, rollouts_per_worker, seed: {}
 
 
-class PipelineStage(object):
-    """A single stage in a training pipeline.
+class TrainingSettings(object):
+    """Class defining parameters used for training (within a stage or the
+    entire pipeline).
+
+    # Attributes
+
+    num_mini_batch : The number of mini-batches to break a rollout into.
+    update_repeats : The number of times we will cycle through the mini-batches corresponding
+        to a single rollout doing gradient updates.
+    max_grad_norm : The maximum "inf" norm of any gradient step (gradients are clipped to not exceed this).
+    num_steps : Total number of steps a single agent takes in a rollout.
+    gamma : Discount factor applied to rewards (should be in [0, 1]).
+    use_gae : Whether or not to use generalized advantage estimation (GAE).
+    gae_lambda : The additional parameter used in GAE.
+    advance_scene_rollout_period: Optional number of rollouts before enforcing an advance scene in all samplers.
+    save_interval : The frequency with which to save (in total agent steps taken). If `None` then *no*
+        checkpoints will be saved. Otherwise, in addition to the checkpoints being saved every
+        `save_interval` steps, a checkpoint will *always* be saved at the end of each pipeline stage.
+        If `save_interval <= 0` then checkpoints will only be saved at the end of each pipeline stage.
+    metric_accumulate_interval : The frequency with which training/validation metrics are accumulated
+        (in total agent steps). Metrics accumulated in an interval are logged (if `should_log` is `True`)
+        and used by the stage's early stopping criterion (if any).
+    """
+
+    # noinspection PyUnresolvedReferences
+    def __init__(
+        self,
+        num_mini_batch: Optional[int] = None,
+        update_repeats: Optional[int] = None,
+        max_grad_norm: Optional[float] = None,
+        num_steps: Optional[int] = None,
+        gamma: Optional[float] = None,
+        use_gae: Optional[bool] = None,
+        gae_lambda: Optional[float] = None,
+        advance_scene_rollout_period: Optional[int] = None,
+        save_interval: Optional[int] = None,
+        metric_accumulate_interval: Optional[int] = None,
+        **kwargs: Any,
+    ):
+        all_vars = prepare_locals_for_super(locals(), ignore_kwargs=True)
+
+        for key, value in all_vars.items():
+            setattr(self, key, value)
```

Comment:
The one unfortunate thing here is that we're not going to get any auto-completion. Can we maybe do something like:
```python
class TrainingSettings(object):
    """...docstring..."
    num_mini_batch: Optional[int]
    update_repeats: Optional[int]
    max_grad_norm: Optional[float]
    num_steps: Optional[int]
    gamma: Optional[float]
    use_gae: Optional[bool]
    gae_lambda: Optional[float]
    advance_scene_rollout_period: Optional[int]
    save_interval: Optional[int]
    metric_accumulate_interval: Optional[int]
    
    def __init__(...):
        ...
```
?

Reply:
I think it's unavoidable if we want to be IDE-friendly. Thanks for the note! 👍